### PR TITLE
Fix VM creation and update when `sshAccess` is disabled.

### PR DESCRIPTION
**Pull request by AleksandarSavchev**
**What this PR does / why we need it**:
This PR fixes creation and update of azure VMs when `sshAccess` is disabled by adding a "fake" `PublicKey`.

Currently, creating an azure shoot with `sshAccess` disabled is not possible since the shoot would fail in reconciliation. It fails with the following cloud provider validation error: `The value of parameter linuxConfiguration.ssh.publicKeys.keyData is invalid.` since its value is "".

After removing `SSH` from `LinuxConfiguration` it starts failing with this cloud provider error: `Authentication using either SSH or by user name and password must be enabled in Linux profile.`, which would suggest we need to provide either `SSH` or an `adminPassword`. We do not include the `adminPassword` and it is ill-advised to do so. The only option left would be to pass a "fake" `PublicKey`.

**Which issue(s) this PR fixes**:
Fixes #

**Special notes for your reviewer**:

**Release note**:

```noteworthy user
Fixed VM creation and update when `sshAccess` is disabled.
```

### CODE REVIEW

**Comment by himanshu-kun** [comment]
@AleksandarSavchev could you mention why an SSH access disablement is required ? Any gardener issue , slack thread ? We would like to know the context , as this will affect all our mcm-providers.
cc @ashwani2k @unmarshall

**Comment by AleksandarSavchev** [comment]
@himanshu-kun disabling SSH access is needed to make our clusters compliant with DISA STIGs rules [242393](https://www.stigviewer.com/stig/kubernetes/2022-12-02/finding/V-242393) and [242394](https://www.stigviewer.com/stig/kubernetes/2022-12-02/finding/V-242394). You can find the backlog issue for these changes [here](https://github.com/gardener/backlog/issues/49). For more information you can check the PR that added `SSHAccess` field to shoot description: [ref](https://github.com/gardener/gardener/pull/7188).

**Comment by unmarshall** [comment]
As per https://github.com/MicrosoftDocs/azure-docs/issues/14447 it should be possible to block SSH access by configuring network security group. Have you already tried this out?

**Comment by kon-angelo** [comment]
@unmarshall there are 2 issues: 
a) The links that @AleksandarSavchev posted show that the endgoal is to disable the sshd daemon of the nodes - not disable SSH access somehow e.g. with NSGs. 
b) If when shoot.spec.sshAccess is disabled, gardener does not create any SSH public key, then the machineclass will be invalid anyway. I think this was the original problem, that the since there isn't a key anymore, machine creation fails anyway.

**Review by himanshu-kun — CHANGES_REQUESTED** [review]
we already have a test for `admin` name  https://github.com/gardener/machine-controller-manager-provider-azure/blob/1d20246175bf6b04d80a3161ba3fd7a972422ccd/pkg/azure/core_test.go#L409-L422

Similarly , pls also add the following test case:
* when `providerSpec.Properties.OsProfile.LinuxConfiguration.SSH.PublicKeys.KeyData` is not supplied , we get the following error

`
Failure sending request: StatusCode=400 -- Original Error: Code="InvalidParameter" Message="The value of parameter linuxConfiguration.ssh.publicKeys.keyData is invalid." Target="linuxConfiguration.ssh.publicKeys.keyData"
`

**Comment by himanshu-kun** [comment]
> When a shoot is configured with `SSHAccess` disabled a service called `sshd-ensurer` runs a script every 30 seconds which disables and stops the `sshd` service if enabled and/or running, it also kills any ssh connections. With `SSHAccess` disabled no public keys are created and the creation of bastions is disabled as well(ref: [gardener/gardener#7188](https://github.com/gardener/gardener/pull/7188))
> 
> The problem of disabling ssh access has already been solved by the upper mentioned functionality. The problem in `provider-azure` is that we cannot create shoots with `SSHAccess` disabled, since there are no SSH public keys.

thanks for the context and sorry for the delay. I tried this out myself ..
the machineClass looked like this when `workersSettings.sshAccess.enabled: false`
```
osProfile:
      adminUsername: core
      linuxConfiguration:
        disablePasswordAuthentication: true
        ssh:
          publicKeys:
            keyData: null
            path: /home/core/.ssh/authorized_keys
```
the error from provider was
```
Cloud provider message - machine codes error: code = [Internal] message = [compute.VirtualMachinesClient#CreateOrUpdate: Failure sending request: StatusCode=400 -- Original Error: Code="InvalidParameter" Message="The value of parameter linuxConfiguration.ssh.publicKeys.keyData is invalid." Target="linuxConfiguration.ssh.publicKeys.keyData"]
```

I have reviewed the PR, kindly address , then we'll merge it.

File: pkg/azure/utils.go [hunk 1]
```diff
@@ -504,6 +533,15 @@ func (d *MachinePlugin) createVMNicDisk(req *driver.CreateMachineRequest) (*comp
 		vmImageRef = &vmImage
 	}
 
+	if len(providerSpec.Properties.OsProfile.LinuxConfiguration.SSH.PublicKeys.KeyData) == 0 {
+		publicKey, err := generatePublicKey()
+		if err != nil {
+			return nil, err
+		}
+
+		providerSpec.Properties.OsProfile.LinuxConfiguration.SSH.PublicKeys.KeyData = publicKey
```
Comment (himanshu-kun) [inline pkg/azure/utils.go:542]: pls print a log of V(3) to inform that a fake ssh key is set to allow VM creation as key not supplied.

**Review by himanshu-kun — APPROVED** [review]
/lgtm
Thnx , this looks better!

**Pull request merged**